File: apps/data-engine/src/contracts/registry.ts
/**
 * Contract registry — loads JSON schemas from the single source of truth
 * at project root `contracts/` and provides typed validators.
 *
 * IMPORTANT: schemas are loaded at module init time and cached. This is
 * acceptable because contracts change only across deployments, not at runtime.
 */

import candleSchema from "../../../../contracts/market-candles.schema.json"
import signalSchema from "../../../../contracts/signals-trading.schema.json"
import orderSchema from "../../../../contracts/orders-execution.schema.json"
import fillSchema from "../../../../contracts/fills-execution.schema.json"
import heartbeatSchema from "../../../../contracts/system-heartbeat.schema.json"
import errorSchema from "../../../../contracts/system-errors.schema.json"

// ── 1. Field-level type info ──────────────────────────────────────

type FieldType = "string" | "number" | "integer" | "boolean" | "object"
type FieldDef = {
  type: FieldType
  required: boolean
  pattern?: string
  enum?: readonly string[]
  min?: number
  max?: number
  const?: number | string | boolean
  default?: unknown
  description?: string
}
type SchemaDef = {
  stream: string
  version: number
  fields: Record<string, FieldDef>
}

// ── 2. Load & validate schema structure ───────────────────────────

function loadSchema(raw: unknown, name: string): SchemaDef {
  const s = raw as Record<string, unknown>
  if (!s.stream || typeof s.stream !== "string") throw new Error(`${name}: missing stream`)
  if (!s.fields || typeof s.fields !== "object") throw new Error(`${name}: missing fields`)
  return s as unknown as SchemaDef
}

const SCHEMAS: Record<string, SchemaDef> = {
  candle:    loadSchema(candleSchema, "market-candles"),
  signal:    loadSchema(signalSchema, "signals-trading"),
  order:     loadSchema(orderSchema, "orders-execution"),
  fill:      loadSchema(fillSchema, "fills-execution"),
  heartbeat: loadSchema(heartbeatSchema, "system-heartbeat"),
  error:     loadSchema(errorSchema, "system-errors"),
}

// ── 3. Validator engine ───────────────────────────────────────────

export interface ValidationError {
  field: string
  rule: string
  expected: string
  actual: string
}

export type ValidationResult =
  | { ok: true }
  | { ok: false; errors: ValidationError[] }

function validateValue(
  value: unknown,
  def: FieldDef,
  fieldPath: string,
): ValidationError | null {
  // -- optional with default --------------------------------------------------
  if (value === undefined || value === null) {
    if (def.required) {
      return { field: fieldPath, rule: "required", expected: "present", actual: String(value) }
    }
    return null
  }

  // -- type -------------------------------------------------------------------
  const actualType = typeof value
  if (def.type === "integer") {
    if (!Number.isInteger(value)) {
      return { field: fieldPath, rule: "type", expected: "integer", actual: actualType }
    }
  } else if (def.type === "number") {
    if (actualType !== "number") {
      return { field: fieldPath, rule: "type", expected: "number", actual: actualType }
    }
  } else if (actualType !== def.type) {
    return { field: fieldPath, rule: "type", expected: def.type, actual: actualType }
  }

  // -- const ------------------------------------------------------------------
  if (def.const !== undefined && value !== def.const) {
    return { field: fieldPath, rule: "const", expected: String(def.const), actual: String(value) }
  }

  // -- enum -------------------------------------------------------------------
  if (def.enum && !def.enum.includes(value as string)) {
    return { field: fieldPath, rule: "enum", expected: def.enum.join("|"), actual: String(value) }
  }

  // -- pattern ----------------------------------------------------------------
  if (def.pattern && typeof value === "string") {
    const re = new RegExp(def.pattern)
    if (!re.test(value)) {
      return { field: fieldPath, rule: "pattern", expected: def.pattern, actual: value }
    }
  }

  // -- min/max ----------------------------------------------------------------
  if (def.min !== undefined && typeof value === "number" && value < def.min) {
    return { field: fieldPath, rule: "min", expected: `>=${def.min}`, actual: String(value) }
  }
  if (def.max !== undefined && typeof value === "number" && value > def.max) {
    return { field: fieldPath, rule: "max", expected: `<=${def.max}`, actual: String(value) }
  }

  return null
}

export function validatePayload(schemaKey: string, payload: Record<string, unknown>): ValidationResult {
  const schema = SCHEMAS[schemaKey]
  if (!schema) {
    return { ok: false, errors: [{ field: "_schema", rule: "unknown", expected: `known key (${Object.keys(SCHEMAS).join(", ")})`, actual: schemaKey }] }
  }

  const errors: ValidationError[] = []

  for (const [fieldName, def] of Object.entries(schema.fields)) {
    const err = validateValue(payload[fieldName], def, fieldName)
    if (err) errors.push(err)
  }

  return errors.length === 0 ? { ok: true } : { ok: false, errors }
}

// ── 4. Typed helpers ──────────────────────────────────────────────

export interface CandlePayload {
  symbol: string
  timeframe: string
  open: string
  high: string
  low: string
  close: string
  volume: string
  timestamp: number
  is_complete?: boolean
  schema_version: number
}

export interface SignalPayload {
  signal_id: string
  symbol: string
  action: "BUY" | "SELL"
  confidence: number
  model_version?: string
  regime?: "TRENDING" | "RANGING" | "UNKNOWN"
  timestamp: number
  schema_version: number
}

export interface OrderPayload {
  order_id: string
  signal_id: string
  symbol: string
  side: "BUY" | "SELL"
  order_type: "MARKET"
  amount: string
  idempotency_key: string
  timestamp: number
  schema_version: number
}

export interface FillPayload {
  fill_id: string
  order_id: string
  symbol: string
  side: "BUY" | "SELL"
  filled_qty: string
  avg_price: string
  status: "FILLED" | "PARTIALLY_FILLED" | "REJECTED"
  exchange_order_id?: string
  error_message?: string
  timestamp: number
  schema_version: number
}

export interface HeartbeatPayload {
  service: "data-engine" | "analytics-engine"
  status: "healthy" | "degraded" | "halted"
  uptime_seconds: number
  mode: "BACKTESTING" | "PAPER_TRADING" | "LIVE" | "LIVE_SIMULATION"
  broker_ok: boolean
  exchange_ok: boolean
  loops_alive?: number
  timestamp: number
  schema_version: number
}

export interface ErrorPayload {
  error_id: string
  service: "data-engine" | "analytics-engine"
  severity: "WARN" | "ERROR" | "CRITICAL"
  error_code: string
  message: string
  metadata?: Record<string, unknown>
  timestamp: number
  schema_version: number
}

export function validateCandle(p: Record<string, unknown>): ValidationResult {
  return validatePayload("candle", p)
}
export function validateSignal(p: Record<string, unknown>): ValidationResult {
  return validatePayload("signal", p)
}
export function validateOrder(p: Record<string, unknown>): ValidationResult {
  return validatePayload("order", p)
}
export function validateFill(p: Record<string, unknown>): ValidationResult {
  return validatePayload("fill", p)
}
export function validateHeartbeat(p: Record<string, unknown>): ValidationResult {
  return validatePayload("heartbeat", p)
}
export function validateError(p: Record<string, unknown>): ValidationResult {
  return validatePayload("error", p)
}

File: apps/analytics-engine/app/contracts/registry.py
"""Contract registry — loads JSON schemas from project root contracts/."""

from __future__ import annotations

import json
import os
import re
from dataclasses import dataclass, field
from typing import Any, Literal


# ── Schema field descriptor ────────────────────────────────────────

@dataclass(frozen=True)
class FieldDef:
    type: str
    required: bool = False
    pattern: str | None = None
    enum: tuple[str, ...] | None = None
    min: float | None = None
    max: float | None = None
    const: int | str | bool | None = None
    default: Any = None
    description: str = ""


@dataclass(frozen=True)
class SchemaDef:
    stream: str
    version: int
    fields: dict[str, FieldDef]


# ── Loader ─────────────────────────────────────────────────────────

_CONTRACTS_DIR = os.path.normpath(
    os.path.join(os.path.dirname(__file__), "..", "..", "..", "..", "contracts")
)

_SOURCES: dict[str, str] = {
    "candle":    "market-candles.schema.json",
    "signal":    "signals-trading.schema.json",
    "order":     "orders-execution.schema.json",
    "fill":      "fills-execution.schema.json",
    "heartbeat": "system-heartbeat.schema.json",
    "error":     "system-errors.schema.json",
}


def _load_raw(name: str) -> dict[str, Any]:
    path = os.path.join(_CONTRACTS_DIR, _SOURCES[name])
    with open(path) as f:
        return json.load(f)


def _parse_field(name: str, raw: dict[str, Any]) -> FieldDef:
    return FieldDef(
        type=raw["type"],
        required=raw.get("required", False),
        pattern=raw.get("pattern"),
        enum=tuple(raw["enum"]) if "enum" in raw else None,
        min=raw.get("min"),
        max=raw.get("max"),
        const=raw.get("const"),
        default=raw.get("default"),
        description=raw.get("description", ""),
    )


_SCHEMAS: dict[str, SchemaDef] = {}

for key in _SOURCES:
    raw = _load_raw(key)
    fields = {fn: _parse_field(fn, fd) for fn, fd in raw["fields"].items()}
    _SCHEMAS[key] = SchemaDef(
        stream=raw["stream"],
        version=raw["version"],
        fields=fields,
    )


# ── Validator ──────────────────────────────────────────────────────

@dataclass
class ValidationError:
    field: str
    rule: str
    expected: str
    actual: str


def _validate_value(value: Any, field_name: str, defn: FieldDef) -> list[ValidationError]:
    errors: list[ValidationError] = []

    # optional with default
    if value is None:
        if defn.required:
            errors.append(ValidationError(field_name, "required", "present", "None"))
        return errors

    # type
    if defn.type == "integer":
        if not isinstance(value, int) or isinstance(value, bool):
            errors.append(ValidationError(field_name, "type", "integer", type(value).__name__))
    elif defn.type == "number":
        if not isinstance(value, (int, float)):
            errors.append(ValidationError(field_name, "type", "number", type(value).__name__))
    elif defn.type == "boolean":
        if not isinstance(value, bool):
            errors.append(ValidationError(field_name, "type", "boolean", type(value).__name__))
    elif defn.type == "object":
        if not isinstance(value, dict):
            errors.append(ValidationError(field_name, "type", "object", type(value).__name__))
    elif defn.type == "string":
        if not isinstance(value, str):
            errors.append(ValidationError(field_name, "type", "string", type(value).__name__))

    # const
    if defn.const is not None and value != defn.const:
        errors.append(ValidationError(field_name, "const", str(defn.const), str(value)))

    # enum
    if defn.enum is not None and value not in defn.enum:
        errors.append(ValidationError(field_name, "enum", "|".join(defn.enum), str(value)))

    # pattern
    if defn.pattern and isinstance(value, str):
        if not re.match(defn.pattern, value):
            errors.append(ValidationError(field_name, "pattern", defn.pattern, value))

    # min / max
    if defn.min is not None and isinstance(value, (int, float)) and value < defn.min:
        errors.append(ValidationError(field_name, "min", f">={defn.min}", str(value)))
    if defn.max is not None and isinstance(value, (int, float)) and value > defn.max:
        errors.append(ValidationError(field_name, "max", f"<={defn.max}", str(value)))

    return errors


def validate(schema_key: str, payload: dict[str, Any]) -> list[ValidationError]:
    """Validate *payload* against the schema identified by *schema_key*.

    Returns an empty list on success.
    """
    schema = _SCHEMAS.get(schema_key)
    if schema is None:
        return [ValidationError("_schema", "unknown", list(_SCHEMAS.keys()).__repr__(), schema_key)]

    errors: list[ValidationError] = []
    for field_name, defn in schema.fields.items():
        errors.extend(_validate_value(payload.get(field_name), field_name, defn))
    return errors


# ── Convenience shortcuts ─────────────────────────────────────────

def validate_candle(p: dict[str, Any]) -> list[ValidationError]:
    return validate("candle", p)

def validate_signal(p: dict[str, Any]) -> list[ValidationError]:
    return validate("signal", p)

def validate_order(p: dict[str, Any]) -> list[ValidationError]:
    return validate("order", p)

def validate_fill(p: dict[str, Any]) -> list[ValidationError]:
    return validate("fill", p)

def validate_heartbeat(p: dict[str, Any]) -> list[ValidationError]:
    return validate("heartbeat", p)

def validate_error(p: dict[str, Any]) -> list[ValidationError]:
    return validate("error", p)
